pulsar-client-cpp/lib/MessageCrypto.h

```diff
@@ -0,0 +1,95 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef LIB_MESSAGECRYPTO_H_
+#define LIB_MESSAGECRYPTO_H_
+
+#include <iostream>
+#include <map>
+#include <set>
+#include <boost/thread/mutex.hpp>
+
+#include <openssl/ssl.h>
+#include <openssl/rand.h>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/engine.h>
+
+#include "SharedBuffer.h"
+#include "ExecutorService.h"
+#include "pulsar/CryptoKeyReader.h"
+#include "PulsarApi.pb.h"
+
+namespace pulsar {
+
+class MessageCrypto {
+ public:
+    typedef std::map<std::string, std::string> StringMap;
+    typedef std::map<std::string, std::pair<std::string, boost::posix_time::ptime>> DataKeyCacheMap;
+
+    MessageCrypto(std::string& logCtx, bool keyGenNeeded);
+    ~MessageCrypto();
+
+    Result addPublicKeyCipher(std::set<std::string>& keyNames, const CryptoKeyReader& keyReader);
+    bool removeKeyCipher(std::string& keyName);
+
+    bool encrypt(std::set<std::string>& encKeys, const CryptoKeyReader& keyReader,
+        proto::MessageMetadata& msgMetadata, SharedBuffer& payload, SharedBuffer& encryptedPayload);
+    bool decrypt(const proto::MessageMetadata& msgMetadata, SharedBuffer& payload,
+        const CryptoKeyReader& keyReader, SharedBuffer& decryptedPayload);
+
+ private:
+
+    typedef boost::unique_lock<boost::mutex> Lock;
+    boost::mutex mutex_;
+
+    int dataKeyLen_;
+    std::unique_ptr<unsigned char []>  dataKey_;
```

Comment:
I think we should stick with `boost::scoped_ptr` instead of `std::unique_ptr` which depends on C++11